FAI-183: Score Cards: Opening a PMML file story

manstis opened the pull request:
See https://issues.redhat.com/browse/FAI-183

This represents the Landing Page when opening (or for now) creating a new PMML file.

When first ran an empty _wrapper_ is shown. It is not part of the editor. It is akin to VSCode without a file opened/created.

![editor-empty](https://user-images.githubusercontent.com/497544/95129126-706f7f00-0752-11eb-8b0f-48f6d0419ce3.png)

When you open a file the list of models will be shown.

![editor-opening-file](https://user-images.githubusercontent.com/497544/95129149-7b2a1400-0752-11eb-8547-8fcaa51f5e87.png)

You can filter the model list by name or supported/unsupported (only Scorecards are supported).

The undo/redo/PMML buttons are not part of the editor. The buttons are provided by the container ("Undo"/"Redo" should be self explanatory, "PMML" gets the XML content of the editor and is used, for example, by VSCode's "Save"). You can delete models and undo/redo the deletion however be aware of https://issues.redhat.com/browse/FAI-270

manstis commented:
@tiagobento IDK if you have any interest in this for now (it's really specific PMML stuff).

Feel free to add yourself as a reviewer if you want to be involved.

manstis commented:
@kelvah I've implemented _atomic design_, in principle, although I've cut it the other way (top level folders are feature, with sub-folders the atomic elements). See what you think.. I can realign with `kogito-apps/ui-packages` if you'd prefer to maintain some form of consistency. @tiagobento the leads over in `kogtio-apps are imposing atomic design.. Roberto has had to follow their mandate over in `kogito-apps` to get a PR approved and merged. We're not lovers of it. Opinion welcome. We've raised the concern.

kelvah reviewed (commented):
Hi @manstis, I left some comments. 

I see your are often declaring variables with a leading `_`. Is that just to silence the name shadowing warning? Are there any other reasons?

I didn't add considerations on file structure because we are discussing it elsewhere, by now you already now my opinion :)

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,121 @@
+                </Title>
+              </TextContent>
+            </SplitItem>
+            <SplitItem isFilled={true}>&nbsp;</SplitItem>
```
Changed as suggested.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,121 @@
+  const [showUnsupportedModels, setShowUnsupportedModels] = useState(true);
+
+  const filterModels = useCallback((): Model[] => {
+    const _lowerCaseFilter: string = filter.toLowerCase();
+    const _filteredModels: Model[] | undefined = models?.filter((_model: Model) => {
+      const _modelName: string | undefined = getModelName(_model);
+      const _isSupportedModelType: boolean = isSupportedModelType(_model);
```
You can't teach an old Java programmer new tricks ;-)

I've removed those you call out here and will endeavour to be a better Type Script citizen!

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,121 @@
+  const filteredModels: Model[] = useMemo(() => filterModels(), [filter, showUnsupportedModels, models]);
+
+  return (
+    <>
```
Removed.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,121 @@
+        <PageSection isFilled={true}>
+          <section>
+            {filteredModels.length > 0 && (
+              <Gallery hasGutter={true}>{filteredModels.map(model => renderCard(model))}</Gallery>
```
OK.. Fair enough. Changed.

manstis commented on packages/pmml-editor/src/editor/index.ts:
```diff
@@ -15,6 +15,7 @@
  */
 
 import "@patternfly/patternfly/patternfly.css";
+import "@patternfly/react-core/dist/styles/base.css";
```
Seems OK. Removed.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/molecules/ActionSelector.tsx:
```diff
@@ -0,0 +1,63 @@
+          id="action-selector-id"
+          data-testid="action-selector__toggle"
+          onToggle={onToggle}
+          toggleIndicator={CaretDownIcon}
```
Done.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/organisms/EmptyStateNoModels.tsx:
```diff
@@ -0,0 +1,37 @@
+ */
+import * as React from "react";
+import { Button, EmptyState, EmptyStateBody, EmptyStateIcon, EmptyStateVariant, Title } from "@patternfly/react-core";
+import BoxesIcon from "@patternfly/react-icons/dist/js/icons/boxes-icon";
```
Fixed all occurrences.

manstis commented on packages/pmml-editor/src/editor/utils/PropertyUtils.ts:
```diff
@@ -0,0 +1,22 @@
+ * limitations under the License.
+ */
+
+export const coalesce = (value: string | undefined | null, defaultValue: string) => {
+  if (value === null || value === undefined) {
+    return defaultValue;
+  }
+  return value;
+};
```
Yes, you are right. I've removed the function and all use. Thank-you.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/molecules/ModelCard.scss:
```diff
@@ -0,0 +1,25 @@
+.model-card {
+  height: 280px;
+
+  &__title {
+    white-space: nowrap;
+    overflow: hidden;
+    text-overflow: ellipsis;
+  }
```
I've removed the `ModelCardTitle` component and added it back to `ModelCard` as suggested above.

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/molecules/ModelCard.tsx:
```diff
@@ -0,0 +1,50 @@
+          <ModelCardIcon model={model} />
+        </CardHeaderMain>
+      </CardHeader>
+      <ModelCardTitle model={model} />
```
Removed. I went to the sub-atomic particle level with my use of atomic design ;-)

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/molecules/ModelCard.tsx:
```diff
@@ -0,0 +1,50 @@
+    <Card data-testid="model-card" isHoverable={true} className="model-card">
+      <CardHeader>
+        <CardHeaderMain>
+          <ModelCardIcon model={model} />
```
I've changed this occurrence (and related code to support `getModelIconByType(type)`).

However I would usually pass a higher level object, when available, in case other properties are needed.

Java Script passes objects by value (but the value is the reference to the object, there is no deep cloning of objects to pass by value, just the pointer is duplicated). Therefore there is no overhead passing a `Model` vs `string` (in fact there is possibly less overhead passing an object reference by value vs a primitive that is cloned).

kelvah replied:
Thank you Michael. I think the issue is more about having a more clear component interface. Passing down individual properties instead of a whole object helps quickly understanding what are the inputs of a component and what's its purpose. It simplifies collaboration when looking at a components tree and eventually it improves debugging.

manstis replied:
Fair point. Did you want more changes on this PR or can it be approved (by you)?

manstis commented on packages/pmml-editor/src/editor/components/LandingPage/organisms/LandingPageToolbar.tsx:
```diff
@@ -0,0 +1,90 @@
+            labelOff="Hide unsupported models"
+            isChecked={showUnsupportedModels}
+            onChange={setShowUnsupportedModels}
+          />
```
Removed different labels (yay) and only render switch when there are unsupported models.

manstis commented:
Hi @kelvah,to answer your first question some variable names (usually those in closures) start with an underscore for the sole purpose of keeping the lint checks happy. Thank-you.

manstis commented:
@tiagobento @jiripetrlik This is ready for your reviews. Thank-you.

tiagobento commented on packages/pmml-editor/src/editor/components/LandingPage/molecules/ActionSelector.tsx:
```diff
@@ -0,0 +1,57 @@
+import { Dropdown, DropdownItem, DropdownPosition, DropdownToggle } from "@patternfly/react-core";
+
+export const ActionSelector = () => {
+  const [isOpen, setOpen] = useState(false);
+
+  const onToggle = (_isOpen: boolean) => {
+    setOpen(_isOpen);
+  };
+  const onSelect = (event: React.SyntheticEvent<HTMLDivElement>) => {
+    setOpen(!isOpen);
+    onFocus();
+  };
+  const onFocus = () => {
+    document.getElementById("action-selector-id")?.focus();
+  };
+
+  const dropdownItems = [
+    <DropdownItem key="createModel" isDisabled={true}>
+      Create model
+    </DropdownItem>,
+    <DropdownItem key="viewDataDictionary" isDisabled={true}>
+      View Data Dictionary
+    </DropdownItem>
+  ];
```
I know there's some debate about whether this is necessary or not, but I'm trying to give people the advice of memoizing in-component functions and values using `useCallback` and `useMemo` respectively. This will prevent us from having unnecessary re-renders. In the case of `dropdownItems`, since it's all static, it can even leave outside the component as a constant.

manstis replied:
Done :-)

tiagobento commented on packages/pmml-editor/src/editor/components/LandingPage/organisms/LandingPageToolbar.tsx:
```diff
@@ -0,0 +1,92 @@
+  const { setFilter, hasUnsupportedModels, showUnsupportedModels, setShowUnsupportedModels } = props;
+
+  const filterField = useRef<HTMLInputElement>(null);
+  const onFilterSubmit = (): void => {
+    if (filterField && filterField.current) {
+      setFilter(filterField.current.value);
+    }
+  };
+  const onFilterEnter = (event: React.KeyboardEvent): void => {
+    if (filterField && filterField.current && event.key === "Enter") {
+      setFilter(filterField.current.value);
+    }
+  };
```
I'm not an expert, but I think if you wrap this input inside a `form`, you're able to have the `Enter` key working out-of-the-box, removing the need for a `ref` and making you write less code. Beware that preventing the default behaviour of the `onSubmit` event is necessary.

manstis replied:
Updated.

tiagobento commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,119 @@
+      const _supportMatch = showUnsupportedModels || _isSupportedModelType;
+      return _nameMatch && _supportMatch;
+    });
+    return _filteredModels ? _filteredModels : [];
```
You can go for a `??` operator here.

manstis replied:
Not the first time I've forgotten about the nullish operator... and probably won't be the last.

Fixed.

tiagobento commented on packages/pmml-editor/src/editor/components/LandingPage/templates/LandingPage.tsx:
```diff
@@ -0,0 +1,119 @@
+                <GalleryItem key={uuid()} data-testid="landing-page__model-card">
+                  <ModelCard
+                    model={model}
+                    onDelete={(_model: Model) => {
+                      dispatch({
+                        type: Actions.DeleteModel,
+                        payload: {
+                          model: _model
+                        }
+                      });
```
Un-memoized function. Might result in unnecessary re-renders.

manstis replied:
Fixed.

tiagobento commented on packages/pmml-editor/src/editor/reducers/PMMLReducer.ts:
```diff
@@ -44,6 +47,18 @@ export const PMMLReducer: HistoryAwareReducer<PMML, AllActions> = (
 
+      case Actions.DeleteModel:
+        return service.mutate(state, null, draft => {
+          const model: Model = action.payload.model;
```
Unnecessary explicit type.

manstis replied:
I must remember this is not Java.... Fixed (and elsewhere).

tiagobento commented on packages/pmml-editor/src/editor/reducers/PMMLReducer.ts:
```diff
@@ -44,6 +47,18 @@ export const PMMLReducer: HistoryAwareReducer<PMML, AllActions> = (
+        return service.mutate(state, null, draft => {
+          const model: Model = action.payload.model;
+          if (draft.models !== undefined) {
+            const models: Model[] = draft.models;
```
Same here.

tiagobento commented on packages/pmml-editor/src/editor/reducers/PMMLReducer.ts:
```diff
@@ -44,6 +47,18 @@ export const PMMLReducer: HistoryAwareReducer<PMML, AllActions> = (
+      case Actions.DeleteModel:
+        return service.mutate(state, null, draft => {
+          const model: Model = action.payload.model;
+          if (draft.models !== undefined) {
```
```suggestion
          if (draft.models) {
```

manstis replied:
Done.

tiagobento commented on packages/pmml-editor/src/editor/reducers/PMMLReducer.ts:
```diff
@@ -44,6 +47,18 @@ export const PMMLReducer: HistoryAwareReducer<PMML, AllActions> = (
+          const model: Model = action.payload.model;
+          if (draft.models !== undefined) {
+            const models: Model[] = draft.models;
+            const index: number = models.indexOf(model);
```
Same here 🙊

manstis replied:
Done.

tiagobento reviewed (commented):
@manstis Thanks for the PR! Left some minor comments inline.

As general comments:
- I'm not a big fan of "utils" or "common" packages, since they tend to aggregate un-related features that are used across the application. This might lead to bad unnecessary coupling and, well.. you know the consequences. If you can, I think it would be better to reorganize things to avoid it.
- I saw that there's some inline texts on the components. I know this PR is not definitive and things will change in the future, but trying to start using the I18n module might be a good idea. Me or @ljmotta can help with that if you need.

manstis commented:
Thank-you for the review @tiagobento.

In response to your general comments:-

> utility classes

I've moved and renamed them to "better" places.

> i18n

We're already tracking the need to use your i18n [here](https://issues.redhat.com/browse/FAI-269).

manstis commented:
Hi @tomasdavidorg IDK if you want to review this instead of/as well as @jiripetrlik.

TBH IDK who from QE has an interest in PMML (editor) UI work.
